clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_while_pn.c

```diff
@@ -1,12 +1,20 @@
 // NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
 // RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve2p1 -S -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
 // RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve2p1 -S -O1 -Werror -Wall -emit-llvm -o - -x c++ %s | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64-none-linux-gnu -target-feature +sve2p1 -S -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64-none-linux-gnu -target-feature +sve2p1 -S -O1 -Werror -Wall -emit-llvm -o - -x c++ %s | FileCheck %s -check-prefix=CPP-CHECK
 // RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve2p1 -S -disable-O0-optnone -Werror -Wall -o /dev/null %s
```

Comment:
Do we need some extra run lines for testing these builtins with sme2 as well?

Reply:
Yes, we do. I'll add it.

Now a questions is do we need a copy of all RUN lines (so we have a set with "+sve2p1" and the same set but with "+sme")
Since the goal is to test this part 
```
let TargetGuard = "sve2p1|sme2 in { ...
```
I would just change a single run line to use "+sme2" and consider that to provide enough test coverage (even though, strictly speaking, someone could move just one line outside that braces).
My problem is that running tests already takes unreasonable amount of time, so I'd rather not make it worse.

Thoughts?

Reply:
Thank you :)

I agree that having a single line to test both sve2p1 and sme2 would be preferable, however I'm not sure if this will be enough coverage when we add the streaming mode attribute to the builtins?

For SME2, we'll need to add `__arm_streaming` to the functions here as the tests should fail without it, but for SVE2p1 we do not need the attribute. @dtemirbulatov recently committed a patch for the pext & psel builtins which has both tests for SME2 & SVE2p1 and passes an extra option for the +sme2 lines to set the attribute: https://github.com/llvm/llvm-project/pull/72827.

Reply:
I meant something else.

Each run line tests all the functions, i.e:
* first line tests all functions, with sve2p1 attribute, compiled as C
* second line tests all functions, with sve2p1 attribute, compiled as C++
* third line tests all functions, short name versions,  with sve2p1 attribute, compiled as C
* fourth line tests all functions, short name versions,  with sve2p1 attribute, compiled as C++

We can surely add four more lines, with sme2 attribute.

But what if we have just:
* first line tests all functions, with **sme2** attribute, compiled as C
* second line tests all functions, with sve2p1 attribute, compiled as C++
* third line tests all functions, short name versions,  with sve2p1 attribute, compiled as C
* fourth line tests all functions, short name versions,  with sve2p1 attribute, compiled as C++

What are we going to lose as test coverage, compared to a variant with eight run lines?

Some odd defect where say, combination of C++ and short/overloaded names misses a check for target feature or
streaming attribute?
But that's *very very likely* not dependent on *this* set of functions, so it's sufficient to have just one function tested with all the combinations and I've already seen such test files.

Reply:
I don't think we need to duplicate all RUN lines for `+sme2`. If we're choosing to use a single RUN line, I would like to suggest this one:

```
RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve2p1 -S -disable-O0-optnone -Werror -Wall -o /dev/null %s
```
(but with `-target-feature +sve -target-feature +sme2` instead of just `-target-feature +sve2p1`)

Because this tests the builtin -> asm flow and ensures we fully support the intrinsics end to end. We can probably trust the lowering from C/C++ -> LLVM IR to happen the same way as it does for SVE2p1, but if we want to have test coverage for that too we could keep one of the other RUN lines you've added in the current version of this patch/PR.

Reply:
Yes, that's what I approximately did in another patch (not yet uploaded):
```
// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sme2 -target-feature +b16b16 -S -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve2 -target-feature +b16b16 -S -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64-none-linux-gnu -target-feature +sve2 -target-feature +b16b16 -S -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64-none-linux-gnu -target-feature +sve2 -target-feature +b16b16 -S -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve2 -target-feature +b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sme2 -target-feature +b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
```

The first  line with `sme2` and an *extra* end-to-end run with `sme2`.
I'll update like this the rest of the in-flight PRs.